File: priv/worker/dtype.py
"""Dtype mapping between nx_tinygrad's stable names, numpy, and tinygrad.

nx_tinygrad owns a small set of stable dtype names that both Elixir and Python
agree on. This module is the single source of truth on the Python side.

v0.1 required: f32, s32, u8. We also map a handful of others that tinygrad and
numpy support natively; unsupported names raise UnsupportedDtype.
"""
from __future__ import annotations

import numpy as np

from errors import UnsupportedDtype

# stable name -> numpy dtype. bf16 has no numpy type, so it uses uint16 as a
# 2-byte transport carrier (the raw bit pattern moves unchanged); the worker
# bitcasts to/from tinygrad's bfloat16 at the boundary.
_NUMPY = {
    "f32": np.float32,
    "f16": np.float16,
    "f64": np.float64,
    "bf16": np.uint16,
    "s8": np.int8,
    "s16": np.int16,
    "s32": np.int32,
    "s64": np.int64,
    "u8": np.uint8,
    "u16": np.uint16,
    "u32": np.uint32,
    "u64": np.uint64,
}

# stable names that are the v0.1 "required" set.
REQUIRED = ("f32", "s32", "u8")


def numpy_dtype(name: str) -> np.dtype:
    try:
        return np.dtype(_NUMPY[name])
    except KeyError:
        raise UnsupportedDtype(f"unsupported dtype: {name!r}") from None


def tinygrad_dtype(name: str):
    """Return the tinygrad DType for a stable name (imported lazily)."""
    from tinygrad import dtypes

    table = {
        "f32": dtypes.float32,
        "f16": dtypes.float16,
        "f64": dtypes.float64,
        "bf16": dtypes.bfloat16,
        "s8": dtypes.int8,
        "s16": dtypes.int16,
        "s32": dtypes.int32,
        "s64": dtypes.int64,
        "u8": dtypes.uint8,
        "u16": dtypes.uint16,
        "u32": dtypes.uint32,
        "u64": dtypes.uint64,
    }
    try:
        return table[name]
    except KeyError:
        raise UnsupportedDtype(f"unsupported dtype: {name!r}") from None


def wire_tensor(arr, name: str, device):
    """Build a tinygrad tensor of wire dtype `name` from a numpy array whose
    dtype is the transport carrier (uint16 for bf16, which is then bitcast)."""
    from tinygrad import Tensor

    t = Tensor(arr, device=device)
    if name == "bf16":
        t = t.bitcast(tinygrad_dtype("bf16"))
    return t


def wire_numpy(tensor, name: str):
    """Convert a tinygrad tensor to a numpy array in the transport carrier
    (bf16 is bitcast to uint16 first, since numpy has no bfloat16)."""
    if name == "bf16":
        from tinygrad import dtypes

        tensor = tensor.bitcast(dtypes.uint16)
    return tensor.numpy()


def itemsize(name: str) -> int:
    return numpy_dtype(name).itemsize


def is_supported(name: str) -> bool:
    return name in _NUMPY

File: lib/nx_tinygrad/dtype.ex
defmodule NxTinygrad.Dtype do
  @moduledoc """
  Mapping between Nx numeric types and nx_tinygrad's stable dtype names.

  The stable names are the wire format shared with the Python worker. Keeping
  the mapping centralized on both sides avoids drift.

  v0.1 required types: `{:f, 32}` <-> `"f32"`, `{:s, 32}` <-> `"s32"`,
  `{:u, 8}` <-> `"u8"`. A few more are mapped for convenience and forward
  compatibility.
  """

  @nx_to_name %{
    {:f, 16} => "f16",
    {:f, 32} => "f32",
    {:f, 64} => "f64",
    {:bf, 16} => "bf16",
    {:s, 8} => "s8",
    {:s, 16} => "s16",
    {:s, 32} => "s32",
    {:s, 64} => "s64",
    {:u, 8} => "u8",
    {:u, 16} => "u16",
    {:u, 32} => "u32",
    {:u, 64} => "u64"
  }

  @name_to_nx Map.new(@nx_to_name, fn {k, v} -> {v, k} end)

  # Types the worker handles today (bf16 rides a uint16 transport carrier and is
  # bitcast to tinygrad bfloat16 in the worker).
  @supported_names ~w(f16 f32 f64 bf16 s8 s16 s32 s64 u8 u16 u32 u64)

  @doc "Stable wire name for an Nx type, or `{:error, reason}`."
  @spec to_name(Nx.Type.t()) :: {:ok, String.t()} | {:error, String.t()}
  def to_name(type) do
    case Map.fetch(@nx_to_name, type) do
      {:ok, name} -> {:ok, name}
      :error -> {:error, "unsupported Nx type: #{inspect(type)}"}
    end
  end

  @doc "Stable wire name for an Nx type, raising `NxTinygrad.CompileError` on failure."
  @spec to_name!(Nx.Type.t()) :: String.t()
  def to_name!(type) do
    case to_name(type) do
      {:ok, name} ->
        name

      {:error, reason} ->
        raise NxTinygrad.CompileError,
          message: reason,
          hint: "supported dtypes: #{Enum.join(@supported_names, ", ")}"
    end
  end

  @doc "Nx type for a stable wire name."
  @spec to_nx(String.t()) :: {:ok, Nx.Type.t()} | {:error, String.t()}
  def to_nx(name) do
    case Map.fetch(@name_to_nx, name) do
      {:ok, type} -> {:ok, type}
      :error -> {:error, "unknown dtype name: #{inspect(name)}"}
    end
  end

  @spec to_nx!(String.t()) :: Nx.Type.t()
  def to_nx!(name) do
    case to_nx(name) do
      {:ok, type} -> type
      {:error, reason} -> raise NxTinygrad.Error, message: reason
    end
  end

  @doc "Whether the stable name is supported by the worker runtime today."
  def supported?(name), do: name in @supported_names

  def supported_names, do: @supported_names
end
